File: packages/android/locanara/src/main/kotlin/com/locanara/builtin/ExtractChain.kt
package com.locanara.builtin

import com.locanara.Entity
import com.locanara.ExtractResult
import com.locanara.composable.Chain
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel

/**
 * Built-in chain for entity extraction.
 *
 * ```kotlin
 * val result = ExtractChain(entityTypes = listOf("person", "location")).run("Tim Cook announced...")
 * println(result.entities.map { it.value })
 * ```
 */
class ExtractChain(
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val entityTypes: List<String> = listOf("person", "location", "date", "organization")
) : Chain {
    override val name = "ExtractChain"

    override suspend fun invoke(input: ChainInput): ChainOutput {
        val prompt = BuiltInPrompts.extract.format(
            mapOf("text" to input.text, "entityTypes" to entityTypes.joinToString(", "))
        )

        println("[ExtractChain] input: ${input.text.take(200)}")
        val response = model.generate(prompt, GenerationConfig.STRUCTURED)
        val text = response.text.trim()
        println("[ExtractChain] output: $text")

        val entities = text.lines()
            .filter { it.isNotBlank() }
            .mapNotNull { line ->
                val trimmed = line.trim()
                val colonIndex = trimmed.indexOf(':')
                if (colonIndex < 0) {
                    Entity(type = "extracted", value = trimmed, confidence = 0.8)
                } else {
                    val type = trimmed.substring(0, colonIndex).trim().lowercase()
                    val value = trimmed.substring(colonIndex + 1).trim()
                    if (value.isEmpty()) null
                    else Entity(type = type, value = value, confidence = 0.9)
                }
            }

        val result = ExtractResult(entities = entities)

        return ChainOutput(
            value = result,
            text = text,
            metadata = input.metadata,
            processingTimeMs = response.processingTimeMs
        )
    }

    /** Type-safe execution that returns [ExtractResult] directly. */
    suspend fun run(text: String): ExtractResult {
        val output = invoke(ChainInput(text = text))
        return output.typed<ExtractResult>()
            ?: throw IllegalStateException("Unexpected output type from ExtractChain")
    }
}

File: packages/apple/Sources/BuiltIn/ExtractChain.swift
import Foundation

/// Built-in chain for entity extraction.
///
/// ```swift
/// let result = try await ExtractChain(entityTypes: ["person", "location"]).run("Tim Cook announced...")
/// print(result.entities.map(\.value))
/// ```
@available(iOS 15.0, macOS 14.0, *)
public struct ExtractChain: Chain {
    public let name = "ExtractChain"

    private let model: any LocanaraModel
    private let entityTypes: [String]

    public init(
        model: (any LocanaraModel)? = nil,
        entityTypes: [String] = ["person", "location", "date", "organization"]
    ) {
        self.model = model ?? LocanaraDefaults.model
        self.entityTypes = entityTypes
    }

    public func invoke(_ input: ChainInput) async throws -> ChainOutput {
        let prompt = try BuiltInPrompts.extract.format([
            "text": input.text,
            "entityTypes": entityTypes.joined(separator: ", ")
        ])

        print("[ExtractChain] input: \(input.text.prefix(200))")
        let response = try await model.generate(prompt: prompt, config: .structured)
        let text = response.text.trimmingCharacters(in: .whitespacesAndNewlines)
        print("[ExtractChain] output: \(text)")

        let entities = text.components(separatedBy: "\n")
            .filter { !$0.isEmpty }
            .compactMap { line -> Entity? in
                let trimmed = line.trimmingCharacters(in: .whitespaces)
                guard let colonIndex = trimmed.firstIndex(of: ":") else {
                    return Entity(type: "extracted", value: trimmed, confidence: 0.8)
                }
                let type = String(trimmed[trimmed.startIndex..<colonIndex]).trimmingCharacters(in: .whitespaces).lowercased()
                let value = String(trimmed[trimmed.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)
                guard !value.isEmpty else { return nil }
                return Entity(type: type, value: value, confidence: 0.9)
            }

        let result = ExtractResult(entities: entities)

        return ChainOutput(
            value: result,
            text: text,
            metadata: input.metadata,
            processingTimeMs: response.processingTimeMs
        )
    }

    /// Type-safe execution that returns `ExtractResult` directly.
    public func run(_ text: String) async throws -> ExtractResult {
        let output = try await invoke(ChainInput(text: text))
        guard let result = output.typed(ExtractResult.self) else {
            throw LocanaraError.executionFailed("Unexpected output type from ExtractChain")
        }
        return result
    }
}
